api/src/main/java/jakarta/servlet/ServletContext.java

```diff
@@ -1436,4 +1459,87 @@
      * @since Servlet 4.0
      */
     public void setResponseCharacterEncoding(String encoding);
+
+    /**
+     * The mode in which this context handles the encoding and decoding of URI paths. This mode affects the behavior of:
+     * <ul>
+     * <li>{@link HttpServletRequest#getContextPath()}</li>
+     * <li>{@link HttpServletRequest#getServletPath()}</li>
+     * <li>{@link HttpServletRequest#getPathInfo()}</li>
+     * <li>{@link HttpServletRequest#getPath()}</li>
+     * <li>{@link HttpServletRequest#getRealPath(String)}</li>
+     * <li>{@link HttpServletRequest#getRequestDispatcher(String)}</li>
+     * <li>{@link ServletContext#getContextPath()}</li>
+     * <li>{@link ServletContext#getContext(String)}</li>
+     * <li>{@link ServletContext#getRequestDispatcher(String)}</li>
+     * <li>{@link ServletContext#getRealPath(String)}</li>
+     * <li>{@link ServletContext#getPath(String)}</li>
+     * <li>{@link ServletContext#getResource(String)}</li>
+     * <li>{@link ServletContext#getResourceAsStream(String)}</li>
+     * <li>{@link ServletContext#getResourceAsPath(String)}</li>
+     * <li>{@link ServletContext#getResourcePaths(String)}</li>
+     * <li>{@link AsyncContext#dispatch(String)}</li>
+     * <li>{@link AsyncContext#dispatch(ServletContext, String)}</li>
+     * </ul>
+     * for the purposes of these modes:
+     * <ul>
+     * <li>a 'reserved character' is a character defined as <code>reserved</code> by
+     * <a href="https://datatracker.ietf.org/doc/html/rfc3986#section-2.2">RFC3986 Section 2.2</a> or the ';'
+     * character.</li>
+     * <li>a 'path parameter' is the part of a URI segment as defined by the obsoleted
+     * <a href="https://datatracker.ietf.org/doc/html/rfc2396#section-3.3">RFC2396 Section 3.3</a>.</li>
+     * </ul>
+     */
+    public enum PathCodingMode {
```

Comment:
The approach I landed on is to have the 3 modes:   `LEGACY` is what we have now and will be the default for any web.xml <= 5.0. The `DECODED` mode is more or less what we've called `SAFE` before, it will be the default for >=6.0 and paths will not have any encodings in them and encoded reserved characters are just illegal.  `ENCODED` is the optional mode that allows encoded reserved characters.

I might be persuaded that the 6.0 default should be `ENCODED`... in which case perhaps we could drop `DECODED` and go to a boolean mode? But naming is difficult and I can't convince myself 100% of this.

I've included path parameters behaviours, but perhaps that needs a n independent mode?

We could put something in web.xml for this... but let's finalize the design before thinking about that.

Reply:
I'd like to get to boolean approach too.... but so far I've been unable to reduce down the behaviours to just two.   So let's iterate on how we see the various modes working... and if we get down to 2 of them, we can go to the binary model and get rid of this enum and associated setter.